### PR TITLE
CGManifest - add training entries and generate entries for submodules.

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Add cgmanifest.json entries for training dependencies.
Add script to generate git submodule cgmanifest.json entries.

**Motivation and Context**
Compliance.

### CODE REVIEW

**Comment by snnn** [comment]
Please do it recursively, for everything onnxruntime may link to.

File: cgmanifest.json [hunk 1]
```diff
@@ -464,7 +464,34 @@
...
+               "commitHash": "373eb09e4c5d2b3cc2493f0949dc4be6b6a45e81",
+               "repositoryUrl": "https://github.com/tensorflow/tensorboard.git"
+            },
+            "type": "git"
+         }
       }
    ],
    "Version": 1
-}
+}
```
Comment (thiagocrepaldi) [inline cgmanifest.json:497]: Missing end of line

**Comment by snnn** [comment]
Things I found:

1. horovod 
2. nccl
3. mpi
4. lbfgs
5. boost
6. flatbuffers

For any of them, even if it already exists in the cgmanifest.json, you may still need to do the register for your own if you need a different version of the software.

The cgmanifest.json has two major purpose:
1. Legal review
2. Security Scan

For the second one, if you didn't provide the correct version number, the security team can not do the job for you.

File: cgmanifests/README.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
+This directory contains Component Governance cgmanifest.json files.
+See here for more details: https://docs.opensource.microsoft.com/tools/cg/cgmanifest.html
+
+Any cgmanifest file entries that don't belong in more specific categories
+(e.g., git submodules) can go into `cgmanifests/cgmanifest.json`.
+
+### Git Submodules
```
Comment (thiagocrepaldi) [inline cgmanifests/README.md:7]: Not usual to have `###` and not `#` and `##` before, as the headings are hierarchical

**Comment by edgchen1** [comment]
generating entries from submodules recursively

---
In reply to: [628340806](https://github.com/microsoft/onnxruntime/pull/3933#issuecomment-628340806) [](ancestors = 628340806)

**Comment by edgchen1** [comment]
for legal and security purposes.
[https://docs.opensource.microsoft.com/tools/cg/index.html](https://docs.opensource.microsoft.com/tools/cg/index.html)

---
In reply to: [411408892](https://github.com/microsoft/onnxruntime/pull/3933#issuecomment-411408892) [](ancestors = 411408892)

**Comment by edgchen1** [comment]
added nccl and mpi (with information from orttraining/readme.md). the other items are among the submodules so should be included in the generated submodule cgmanifest.json. @SherlockNoMad, @manashgoswami, am i missing anything else?

---
In reply to: [628822553](https://github.com/microsoft/onnxruntime/pull/3933#issuecomment-628822553) [](ancestors = 628822553)

File: cgmanifests/submodules/cgmanifest.json [hunk 3]
```diff
@@ -0,0 +1,655 @@
...
+      }
+    },
+    {
+      "component": {
+        "type": "git",
+        "git": {
+          "commitHash": "554d8f538e4b0004152c75a9eb72d99f8c0f914c",
+          "repositoryUrl": "https://github.com/eigenteam/eigen-git-mirror.git"
+        },
+        "comments": "git submodule at cmake/external/horovod/third_party/eigen"
```
Comment (snnn) [inline cgmanifests/submodules/cgmanifest.json:381]: Is it possible to remove some unused components from here?
Like, the eigen from github. We're not using it.

**Comment by snnn** [comment]
Please also update 
https://github.com/microsoft/onnxruntime/blob/master/docs/How_To_Update_ONNX_Dev_Notes.md

File: cgmanifests/submodules/cgmanifest.json (hunk 3 above)
Comment (edgchen1) [reply cgmanifests/submodules/cgmanifest.json:381]: It is possible. However, we are still cloning whatever submodules are here. If it doesn't hurt listing them here I'd rather leave them. If we ignore some, we need to keep track of which ones we are ignoring and make sure to not inadvertently use them in the future.

---
In reply to: [425492748](https://github.com/microsoft/onnxruntime/pull/3933#discussion_r425492748) [](ancestors = 425492748)

Comment (edgchen1) [reply cgmanifests/submodules/cgmanifest.json:381]: Regarding this eigen in particular, I believe there are plans to remove the horovod dependency. When that happens, this should go away too.

---
In reply to: [425492748](https://github.com/microsoft/onnxruntime/pull/3933#discussion_r425492748) [](ancestors = 425492748)

Comment (snnn) [reply cgmanifests/submodules/cgmanifest.json:381]: "we need to keep track of which ones we are ignoring"

We may need to do it in the future. Unused submodules can generate unnecessary security warnings.

Comment (edgchen1) [reply cgmanifests/submodules/cgmanifest.json:381]: sounds good. it should be straightforward to check a list of paths to ignore in the generate script.

**Pull request merged**